lib/modules/manager/gradle/parser/common.ts

```diff
@@ -0,0 +1,101 @@
+import type { lexer } from 'good-enough-parser';
+import type { Ctx, NonEmptyArray, PackageVariables } from '../types';
+
+export const REGISTRY_URLS = {
+  google: 'https://dl.google.com/android/maven2/',
+  gradlePluginPortal: 'https://plugins.gradle.org/m2/',
+  jcenter: 'https://jcenter.bintray.com/',
+  mavenCentral: 'https://repo.maven.apache.org/maven2',
+};
+
+export const ANNOYING_METHODS: ReadonlySet<string> = new Set([
+  'createXmlValueRemover',
+  'events',
+  'args',
+  'arrayOf',
+  'listOf',
+  'mutableListOf',
+  'setOf',
+  'mutableSetOf',
+  'stages', // https://github.com/ajoberstar/reckon,
+  'mapScalar', // https://github.com/apollographql/apollo-kotlin
+]);
+
+export function storeVarToken(ctx: Ctx, node: lexer.Token): Ctx {
+  ctx.varTokens.push(node);
+  return ctx;
+}
+
+export function storeInTokenMap(ctx: Ctx, tokenMapKey: string): Ctx {
+  ctx.tokenMap[tokenMapKey] = ctx.varTokens;
+  ctx.varTokens = [];
+
+  return ctx;
+}
+
+export function loadFromTokenMap(
+  ctx: Ctx,
+  tokenMapKey: string
+): NonEmptyArray<lexer.Token> {
+  const tokens = ctx.tokenMap[tokenMapKey];
+  if (!tokens) {
+    throw new Error(`Expected token ${tokenMapKey} not found`);
+  }
+
+  return tokens as NonEmptyArray<lexer.Token>;
+}
+
+export function cleanupTempVars(ctx: Ctx): Ctx {
+  ctx.tokenMap = {};
+  ctx.varTokens = [];
+
+  return ctx;
+}
+
+export function stripReservedPrefixFromKeyTokens(ctx: Ctx): Ctx {
+  const unwantedPrefixes = ['ext', 'extra', 'project', 'rootProject'];
+  while (
+    ctx.varTokens.length > 1 && // ensures there will be always at least one token
+    ctx.varTokens[0] &&
+    unwantedPrefixes.includes(ctx.varTokens[0].value)
+  ) {
+    ctx.varTokens.shift();
+  }
+
+  return ctx;
+}
+
+export function coalesceVariable(ctx: Ctx): Ctx {
+  if (ctx.varTokens.length > 1) {
+    ctx.varTokens[0]!.value = ctx.varTokens
+      .map((token) => token.value)
+      .join('.');
+    ctx.varTokens.length = 1;
+  }
+
+  return ctx;
+}
+
+export function interpolateString(
+  childTokens: lexer.Token[],
+  variables: PackageVariables
+): string | null {
+  const resolvedSubstrings: string[] = [];
+  for (const childToken of childTokens) {
+    const type = childToken.type;
+    if (type === 'string-value') {
+      resolvedSubstrings.push(childToken.value);
+    } else if (type === 'symbol') {
+      const varData = variables[childToken.value];
+      if (varData) {
+        resolvedSubstrings.push(varData.value);
+      } else {
+        return null;
+      }
+    } else {
+      return null;
+    }
```

Comment:
you could simplify the if/else with:
 
if (type !== 'string-value' || type !== 'symbol') {
return null
}

then check if/else with the types.

same thing for varData.

it could be if(!varData) {return null}
then push  varData.value

exit as soon as possible, makes code cleaner

Reply:
altho this might be a copy/paste from legacy code, so feel free to keep it

Reply:
Correct would be: `if (type !== 'string-value' && type !== 'symbol') {`
The number of branches would be the same, so I don't really see the value of changing this, to be honest.

I've taken this method almost verbatim (only token types were changed) from the existing implementation:

https://github.com/renovatebot/renovate/blob/c0b15fb8055714660a522d206b71816a218cf93b/lib/modules/manager/gradle/utils.ts#L86-L108